// react-md/lib/DataTables/EditDialogColumn.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'react-dom', 'classnames', '../constants/keyCodes', '../constants/CSSTransitionGroupTick', '../utils/getField', '../utils/PropTypes/invalidIf', '../Dialogs/DialogFooter', './TableColumn', '../TextFields/TextField', '../FontIcons/FontIcon'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('react-dom'), require('classnames'), require('../constants/keyCodes'), require('../constants/CSSTransitionGroupTick'), require('../utils/getField'), require('../utils/PropTypes/invalidIf'), require('../Dialogs/DialogFooter'), require('./TableColumn'), require('../TextFields/TextField'), require('../FontIcons/FontIcon'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactDom, global.classnames, global.keyCodes, global.CSSTransitionGroupTick, global.getField, global.invalidIf, global.DialogFooter, global.TableColumn, global.TextField, global.FontIcon);
    global.EditDialogColumn = mod.exports;
  }
})(this, function (exports, _react, _reactDom, _classnames, _keyCodes, _CSSTransitionGroupTick, _getField, _invalidIf, _DialogFooter, _TableColumn, _TextField, _FontIcon) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  var _CSSTransitionGroupTick2 = _interopRequireDefault(_CSSTransitionGroupTick);

  var _getField2 = _interopRequireDefault(_getField);

  var _invalidIf2 = _interopRequireDefault(_invalidIf);

  var _DialogFooter2 = _interopRequireDefault(_DialogFooter);

  var _TableColumn2 = _interopRequireDefault(_TableColumn);

  var _TextField2 = _interopRequireDefault(_TextField);

  var _FontIcon2 = _interopRequireDefault(_FontIcon);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  function _objectWithoutProperties(obj, keys) {
    var target = {};

    for (var i in obj) {
      if (keys.indexOf(i) >= 0) continue;
      if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
      target[i] = obj[i];
    }

    return target;
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var EditDialogColumn = function (_PureComponent) {
    _inherits(EditDialogColumn, _PureComponent);

    function EditDialogColumn(props, context) {
      _classCallCheck(this, EditDialogColumn);

      var _this = _possibleConstructorReturn(this, (EditDialogColumn.__proto__ || Object.getPrototypeOf(EditDialogColumn)).call(this, props, context));

      _this.state = {
        value: props.defaultValue,
        active: false
      };

      _this._setColumn = _this._setColumn.bind(_this);
      _this._setField = _this._setField.bind(_this);
      _this._setOkButton = _this._setOkButton.bind(_this);
      _this._save = _this._save.bind(_this);
      _this._overrideTab = _this._overrideTab.bind(_this);
      _this._handleClick = _this._handleClick.bind(_this);
      _this._handleKeyUp = _this._handleKeyUp.bind(_this);
      _this._handleKeyDown = _this._handleKeyDown.bind(_this);
      _this._handleChange = _this._handleChange.bind(_this);
      _this._handleTouchStart = _this._handleTouchStart.bind(_this);
      _this._handleTouchEnd = _this._handleTouchEnd.bind(_this);
      _this._handleCancelClick = _this._handleCancelClick.bind(_this);
      _this._handleClickOutside = _this._handleClickOutside.bind(_this);
      _this._handleMouseOver = _this._handleMouseOver.bind(_this);
      _this._handleMouseLeave = _this._handleMouseLeave.bind(_this);
      return _this;
    }

    _createClass(EditDialogColumn, [{
      key: 'componentWillUpdate',
      value: function componentWillUpdate(nextProps, nextState) {
        var _this2 = this;

        if (this.state.active === nextState.active) {
          return;
        }

        if (nextState.active) {
          // Wait for a re-render cycle before adding the window click event.
          // This will be called immediately after being clicked open and close
          // the dialog immediately on open.
          this._clickTimeout = setTimeout(function () {
            _this2._clickTimeout = null;
            window.addEventListener('click', _this2._handleClickOutside);
          }, _CSSTransitionGroupTick2.default);
        } else {
          if (this._clickTimeout) {
            clearTimeout(this._clickTimeout);
            this._clickTimeout = null;
          }
          window.removeEventListener('click', this._handleClickOutside);
        }

        if (this._timeout) {
          clearTimeout(this._timeout);
        }

        this._timeout = setTimeout(function () {
          if (!nextState.active && _this2._field) {
            _this2._field.blur();
          }
          _this2._timeout = null;
          _this2.setState({ animating: false });
        }, nextProps.transitionDuration);
        this.setState({ animating: true });
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        window.removeEventListener('click', this._handleClickOutside);

        if (this._timeout) {
          clearTimeout(this._timeout);
        }

        if (this._clickTimeout) {
          clearTimeout(this._clickTimeout);
        }
      }
    }, {
      key: '_setColumn',
      value: function _setColumn(column) {
        this._column = (0, _reactDom.findDOMNode)(column);
      }
    }, {
      key: '_setField',
      value: function _setField(field) {
        if (field) {
          this._field = field.getField();
        }
      }
    }, {
      key: '_setOkButton',
      value: function _setOkButton(okButton) {
        this._okButton = (0, _reactDom.findDOMNode)(okButton);
      }
    }, {
      key: '_handleClickOutside',
      value: function _handleClickOutside(e) {
        if (this._column && !this._column.contains(e.target)) {
          if (this.props.onOutsideClick) {
            this.props.onOutsideClick(e);
          }

          if (this.props.okOnOutsideClick) {
            this._save(e);
          } else {
            this._handleCancelClick(e);
          }
        }
      }
    }, {
      key: '_handleClick',
      value: function _handleClick(e) {
        if (this.props.onClick) {
          this.props.onClick(e);
        }

        if (this.props.inline || this.state.active) {
          return;
        }

        this.setState({ active: true, cancelValue: (0, _getField2.default)(this.props, this.state, 'value') });
      }
    }, {
      key: '_handleKeyUp',
      value: function _handleKeyUp(e) {
        var _this3 = this;

        if (this.props.onKeyUp) {
          this.props.onKeyUp(e);
        }

        // Make sure this is really a _focus_ event from keyboard
        if ((e.which || e.keyCode) !== _keyCodes.TAB || this.state.active || this.props.inline) {
          return;
        }

        // To get a smooth transition with keybaord, need to _emulate_ how the mouse interaction works.
        // Basically position the edit field absolutely, wait for a re-render, then activate the dialog.
        this._timeout = setTimeout(function () {
          _this3._timeout = null;
          _this3.setState({ active: true, cancelValue: (0, _getField2.default)(_this3.props, _this3.state, 'value') });
        }, _CSSTransitionGroupTick2.default);

        this.setState({ absolute: true });
      }
    }, {
      key: '_handleTouchStart',
      value: function _handleTouchStart(e) {
        if (this.props.onTouchStart) {
          this.props.onTouchStart(e);
        }

        this.setState({ absolute: true });
      }
    }, {
      key: '_handleTouchEnd',
      value: function _handleTouchEnd(e) {
        if (this.props.onTouchEnd) {
          this.props.onTouchEnd(e);
        }

        if (this.state.active || this.props.inline) {
          return;
        }

        this.setState({ active: true, cancelValue: (0, _getField2.default)(this.props, this.state, 'value') });
      }
    }, {
      key: '_handleKeyDown',
      value: function _handleKeyDown(e) {
        var onKeyDown = this.props.onKeyDown;

        if (onKeyDown) {
          onKeyDown(e);
        }

        var key = e.which || e.keyCode;
        if (key === _keyCodes.ENTER) {
          this._save(e);
        } else if (key === _keyCodes.TAB) {
          this._overrideTab(e);
        } else if (key === _keyCodes.ESC) {
          this._handleCancelClick(e);
        }
      }
    }, {
      key: '_overrideTab',
      value: function _overrideTab(e) {
        var _props = this.props,
            large = _props.large,
            inline = _props.inline;

        var key = e.which || e.keyCode;
        if (key !== _keyCodes.TAB) {
          return;
        } else if (inline) {
          this._save(e);
          return;
        } else if (!large) {
          if ((0, _getField2.default)(this.props, this.state, 'value')) {
            e.preventDefault();
          } else {
            this.setState({ active: false });
          }
          return;
        }

        var shiftKey = e.shiftKey;
        var classList = e.target.classList;


        var nextFocus = void 0;
        if (classList.contains('md-text-field') && shiftKey) {
          nextFocus = this._okButton;
        } else if (classList.contains('md-btn') && !shiftKey) {
          nextFocus = this._field;
        }

        if (nextFocus) {
          e.preventDefault();
          nextFocus.focus();
        }
      }
    }, {
      key: '_save',
      value: function _save(e) {
        if (this.props.onOkClick) {
          this.props.onOkClick((0, _getField2.default)(this.props, this.state, 'value'), e);
        }

        this.setState({ active: false, absolute: false });
      }
    }, {
      key: '_handleCancelClick',
      value: function _handleCancelClick(e) {
        if (this.props.onCancelClick) {
          this.props.onCancelClick(this.state.cancelValue, e);
        }

        this.setState({ absolute: false, active: false, value: this.state.cancelValue });
      }
    }, {
      key: '_handleChange',
      value: function _handleChange(value, e) {
        if (this.props.onChange) {
          this.props.onChange(value, e);
        }

        if (typeof this.props.value === 'undefined') {
          this.setState({ value: value });
        }
      }
    }, {
      key: '_handleMouseOver',
      value: function _handleMouseOver(e) {
        if (this.props.onMouseOver) {
          this.props.onMouseOver(e);
        }

        if (this.props.inline) {
          return;
        }

        this.setState({ absolute: true });
      }
    }, {
      key: '_handleMouseLeave',
      value: function _handleMouseLeave(e) {
        if (this.props.onMouseLeave) {
          this.props.onMouseLeave(e);
        }

        if (this.props.inline) {
          return;
        }

        this.setState({ absolute: false });
      }
    }, {
      key: 'render',
      value: function render() {
        var rowId = this.context.rowId;
        var _state = this.state,
            active = _state.active,
            absolute = _state.absolute,
            animating = _state.animating;

        var _props2 = this.props,
            style = _props2.style,
            className = _props2.className,
            dialogStyle = _props2.dialogStyle,
            dialogClassName = _props2.dialogClassName,
            maxLength = _props2.maxLength,
            title = _props2.title,
            okLabel = _props2.okLabel,
            cancelLabel = _props2.cancelLabel,
            large = _props2.large,
            label = _props2.label,
            placeholder = _props2.placeholder,
            inline = _props2.inline,
            inlineIconChildren = _props2.inlineIconChildren,
            inlineIconClassName = _props2.inlineIconClassName,
            noIcon = _props2.noIcon,
            header = _props2.header,
            props = _objectWithoutProperties(_props2, ['style', 'className', 'dialogStyle', 'dialogClassName', 'maxLength', 'title', 'okLabel', 'cancelLabel', 'large', 'label', 'placeholder', 'inline', 'inlineIconChildren', 'inlineIconClassName', 'noIcon', 'header']);

        delete props.onMouseOver;
        delete props.onMouseLeave;
        delete props.onTouchStart;
        delete props.onTouchEnd;
        delete props.value;
        delete props.defaultValue;
        delete props.onOkClick;
        delete props.onCancelClick;
        delete props.header;
        delete props.okOnOutsideClick;
        delete props.transitionDuration;

        var value = (0, _getField2.default)(this.props, this.state, 'value');

        var actions = void 0;
        var largeTitle = void 0;
        if (!inline && large && active) {
          actions = [{
            label: cancelLabel,
            onClick: this._handleCancelClick,
            primary: true
          }, {
            label: okLabel,
            onClick: this._save,
            primary: true,
            ref: this._setOkButton,
            onKeyDown: this._overrideTab
          }];

          actions = _react2.default.createElement(_DialogFooter2.default, { actions: actions });

          largeTitle = _react2.default.createElement(
            'h3',
            { className: 'md-title' },
            title
          );
        }

        var pointer = (0, _classnames2.default)({ 'md-pointer--hover': !active });
        var inlineEditIcon = void 0;
        if (inline && !noIcon) {
          inlineEditIcon = _react2.default.createElement(
            _FontIcon2.default,
            {
              key: 'edit-icon',
              style: { marginBottom: 0 },
              iconClassName: inlineIconClassName
            },
            inlineIconChildren
          );
        }

        return _react2.default.createElement(
          _TableColumn2.default,
          {
            style: style,
            className: (0, _classnames2.default)('prevent-grow md-edit-dialog-column', {
              'md-edit-dialog-column--animating': !inline && (absolute || active || animating),
              'md-edit-dialog-column--active': active
            }, className),
            header: header,
            ref: this._setColumn,
            onMouseOver: this._handleMouseOver,
            onMouseLeave: this._handleMouseLeave,
            onClick: this._handleClick,
            onTouchStart: this._handleTouchStart,
            onTouchEnd: this._handleTouchEnd
          },
          _react2.default.createElement(
            'div',
            {
              style: dialogStyle,
              className: (0, _classnames2.default)('md-edit-dialog', {
                'md-edit-dialog--active': active,
                'md-edit-dialog--inline': inline,
                'md-background': active
              }, dialogClassName)
            },
            largeTitle,
            _react2.default.createElement(_TextField2.default, _extends({
              id: rowId + '-edit-dialog'
            }, props, {
              ref: this._setField,
              label: active ? label : null,
              active: active,
              floating: active,
              placeholder: active ? placeholder : placeholder || label,
              block: !active,
              paddedBlock: false,
              className: pointer,
              inputClassName: pointer,
              onKeyUp: this._handleKeyUp,
              onKeyDown: this._handleKeyDown,
              value: value,
              onChange: this._handleChange,
              maxLength: active ? maxLength : null,
              rightIcon: inlineEditIcon
            })),
            actions
          )
        );
      }
    }]);

    return EditDialogColumn;
  }(_react.PureComponent);

  EditDialogColumn.propTypes = {
    /**
     * The optional style to apply to the edit dialog's column.
     */
    style: _react.PropTypes.object,

    /**
     * The optional className to apply to the edit dialog's column.
     */
    className: _react.PropTypes.string,

    /**
     * The optional style to apply to the edit dialog.
     */
    dialogStyle: _react.PropTypes.object,

    /**
     * The optional className to apply to the edit dialog.
     */
    dialogClassName: _react.PropTypes.string,

    /**
     * The transition duration when the dialog is moving from
     * active to inactive.
     */
    transitionDuration: _react.PropTypes.number.isRequired,

    /**
     * Boolean if the edit dialog is disabled.
     */
    disabled: _react.PropTypes.bool,

    /**
     * The optional max length for the edit dialog.
     */
    maxLength: _react.PropTypes.number,

    /**
     * A value to use for the edit dialog text field. This
     * will make the component controlled so you will need
     * to provide an `onChange` function.
     */
    value: _react.PropTypes.string,

    /**
     * An optional function to call when the text field's value
     * is changed. It is called with `(newValue, changeEvent)`.
     */
    onChange: _react.PropTypes.func,

    /**
     * The default value for the column.
     */
    defaultValue: _react.PropTypes.string,

    /**
     * An optional function to call when the input is clicked.
     */
    onClick: _react.PropTypes.func,

    /**
     * An optional function to call when the keyup event is triggered.
     */
    onKeyUp: _react.PropTypes.func,

    /**
     * An optional function to call when the keydown event is triggered.
     */
    onKeyDown: _react.PropTypes.func,

    /**
     * An optional function to call when the mouseover event is triggered.
     */
    onMouseOver: _react.PropTypes.func,

    /**
     * An optional function to call when the mouseleave event is triggered.
     */
    onMouseLeave: _react.PropTypes.func,

    /**
     * An optional function to call when the touchstart event is triggered.
     */
    onTouchStart: _react.PropTypes.func,

    /**
     * An optional function to call when the touchend event is triggered.
     */
    onTouchEnd: _react.PropTypes.func,

    /**
     * Boolean if the edit dialog should be large.
     */
    large: _react.PropTypes.bool,

    /**
     * The title for the large edit dialog. The custom validation changes to required
     * when the `large` prop is set to true.
     */
    title: function title(props, propName, component) {
      for (var _len = arguments.length, others = Array(_len > 3 ? _len - 3 : 0), _key = 3; _key < _len; _key++) {
        others[_key - 3] = arguments[_key];
      }

      if (props.large) {
        var _PropTypes$string;

        return (_PropTypes$string = _react.PropTypes.string).isRequired.apply(_PropTypes$string, [props, propName, component].concat(others));
      } else {
        return _react.PropTypes.string.apply(_react.PropTypes, [props, propName, component].concat(others));
      }
    },

    /**
     * An optional function to call when the OK button is clicked.
     * It is called with `(textFieldValue, clickEvent)`. This function
     * will also be called when a user pressed the enter key.
     */
    onOkClick: _react.PropTypes.func,

    /**
     * The label to use for the OK button.
     */
    okLabel: _react.PropTypes.string.isRequired,

    /**
     * An optional function to call when the Cancel button is clicked.
     * It is called with `(textFieldValueBeforeEdit, clickEvent)`. This
     * function will also be called when the user presses the escape key.
     */
    onCancelClick: _react.PropTypes.func,

    /**
     * The label to use for the Cancel button.
     */
    cancelLabel: _react.PropTypes.string.isRequired,

    /**
     * An optional function to call when the edit dialog is open and the user clicks
     * somewhere else on the page.
     */
    onOutsideClick: _react.PropTypes.func,

    /**
     * A boolean if the action when the edit dialog is open and the user clicks somewhere
     * else on the page should be to confirm the current changes.
     *
     * If this is set to `true`, `onOkClick` will be called. Otherwise `onCancelClick` will
     * be called.
     */
    okOnOutsideClick: _react.PropTypes.bool.isRequired,

    /**
     * An optional label for the text field.
     */
    label: _react.PropTypes.node,

    /**
     * An optional placeholder for the text field.
     */
    placeholder: _react.PropTypes.string,

    /**
     * Boolean if the text field should not appear in a dialog.
     */
    inline: (0, _invalidIf2.default)(_react.PropTypes.bool, 'title', 'large'),

    /**
     * Any children used to display an inline edit dialog's edit icon.
     */
    inlineIconChildren: _react.PropTypes.node,

    /**
     * The icon className used to display the inline edit dialog's edit icon.
     */
    inlineIconClassName: _react.PropTypes.string,

    /**
     * Boolean if an inline edit text field should not include an icon.
     */
    noIcon: (0, _invalidIf2.default)(_react.PropTypes.bool, 'title', 'large'),

    /**
     * This is injected by the `TableRow` component.
     */
    header: _react.PropTypes.bool
  };
  EditDialogColumn.contextTypes = {
    rowId: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]).isRequired
  };
  EditDialogColumn.defaultProps = {
    defaultValue: '',
    transitionDuration: 300,
    okOnOutsideClick: true,
    okLabel: 'Save',
    cancelLabel: 'Cancel',
    inlineIconChildren: 'edit'
  };
  exports.default = EditDialogColumn;
});
