// react-md/lib/DataTables/TableRow.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'react-dom', 'classnames', './headerContextTypes', './rowContextTypes', './TableCheckbox'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('react-dom'), require('classnames'), require('./headerContextTypes'), require('./rowContextTypes'), require('./TableCheckbox'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactDom, global.classnames, global.headerContextTypes, global.rowContextTypes, global.TableCheckbox);
    global.TableRow = mod.exports;
  }
})(this, function (exports, _react, _reactDom, _classnames, _headerContextTypes, _rowContextTypes, _TableCheckbox) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  var _headerContextTypes2 = _interopRequireDefault(_headerContextTypes);

  var _rowContextTypes2 = _interopRequireDefault(_rowContextTypes);

  var _TableCheckbox2 = _interopRequireDefault(_TableCheckbox);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  function _objectWithoutProperties(obj, keys) {
    var target = {};

    for (var i in obj) {
      if (keys.indexOf(i) >= 0) continue;
      if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
      target[i] = obj[i];
    }

    return target;
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var TableRow = function (_Component) {
    _inherits(TableRow, _Component);

    function TableRow(props, context) {
      _classCallCheck(this, TableRow);

      var _this = _possibleConstructorReturn(this, (TableRow.__proto__ || Object.getPrototypeOf(TableRow)).call(this, props, context));

      _this.state = {
        biggest: null,
        widths: [],
        hover: false
      };

      _this._handleMouseOver = _this._handleMouseOver.bind(_this);
      _this._handleMouseLeave = _this._handleMouseLeave.bind(_this);
      _this._setLongestColumn = _this._setLongestColumn.bind(_this);
      return _this;
    }

    _createClass(TableRow, [{
      key: 'getChildContext',
      value: function getChildContext() {
        var _context = this.context,
            baseId = _context.baseId,
            context = _objectWithoutProperties(_context, ['baseId']);

        return _extends({}, context, {
          rowId: context.header ? baseId + 'CheckboxToggleAll' : '' + baseId + this.props.index
        });
      }
    }, {
      key: 'componentDidMount',
      value: function componentDidMount() {
        if (this.props.autoAdjust) {
          this._setLongestColumn();
        }
      }
    }, {
      key: '_ignoreHoverState',
      value: function _ignoreHoverState(classList) {
        return classList.contains('md-list--menu') || ['md-edit-dialog', 'md-edit-dialog--active'].every(function (className) {
          return classList.contains(className);
        });
      }
    }, {
      key: '_handleMouseOver',
      value: function _handleMouseOver(e) {
        if (this.props.onMouseOver) {
          this.props.onMouseOver(e);
        }

        if (this.context.header) {
          return;
        }

        var target = e.target;
        while (target && target.parentNode) {
          if (target.classList && this._ignoreHoverState(target.classList)) {
            this.setState({ hover: false });
            return;
          }

          target = target.parentNode;
        }

        this.setState({ hover: true });
      }
    }, {
      key: '_handleMouseLeave',
      value: function _handleMouseLeave(e) {
        if (this.props.onMouseLeave) {
          this.props.onMouseLeave(e);
        }

        if (this.context.header) {
          return;
        }

        this.setState({ hover: false });
      }
    }, {
      key: '_setLongestColumn',
      value: function _setLongestColumn() {
        var widths = [];
        var biggest = Array.prototype.slice.call((0, _reactDom.findDOMNode)(this).querySelectorAll('.md-table-column')).reduce(function (prev, curr, i) {
          if (curr.classList.contains('prevent-grow')) {
            return prev;
          }

          var width = curr.offsetWidth;
          widths.push(width);
          if (prev.width < width) {
            return { i: i, width: width };
          } else {
            return prev;
          }
        }, { width: 0, i: 0 });

        this.setState({ biggest: biggest, widths: widths });
      }
    }, {
      key: 'render',
      value: function render() {
        var _this2 = this;

        var _state = this.state,
            biggest = _state.biggest,
            widths = _state.widths,
            hover = _state.hover;

        var _props = this.props,
            className = _props.className,
            children = _props.children,
            selected = _props.selected,
            onCheckboxClick = _props.onCheckboxClick,
            props = _objectWithoutProperties(_props, ['className', 'children', 'selected', 'onCheckboxClick']);

        delete props.autoAdjust;
        delete props.index;

        var checkbox = void 0;
        if (!this.context.plain) {
          checkbox = _react2.default.createElement(_TableCheckbox2.default, { key: 'checkbox', checked: selected, onChange: onCheckboxClick });
        }

        var columns = _react.Children.map(children, function (column, i) {
          return _react2.default.cloneElement(column, _extends({}, column.props, {
            header: typeof column.props.header === 'undefined' ? _this2.context.header : column.props.header,
            className: (0, _classnames2.default)({
              'md-table-column--grow': biggest && biggest.i === i,
              // Not last item and the biggest width is greater than this item
              'md-table-column--adjusted': children.length > i + 1 && biggest && biggest.width > widths[i]
            }, column.props.className)
          }));
        });

        return _react2.default.createElement(
          'tr',
          _extends({}, props, {
            className: (0, _classnames2.default)('md-table-row', className, {
              'md-table-row--hover': hover,
              'md-table-row--active': !this.context.header && selected
            }),
            onMouseOver: this._handleMouseOver,
            onMouseLeave: this._handleMouseLeave
          }),
          checkbox,
          columns
        );
      }
    }]);

    return TableRow;
  }(_react.Component);

  TableRow.propTypes = {
    /**
     * An optional style to apply.
     */
    style: _react.PropTypes.object,

    /**
     * An optional className to apply to the row.
     */
    className: _react.PropTypes.string,

    /**
     * A list of `TableColumn` to display in the table.
     *
     * > There should be at least 3 columns in a Data table (non plain)
     */
    children: _react.PropTypes.arrayOf(_react.PropTypes.element).isRequired,

    /**
     * An optional onClick function to call when a row is clicked.
     */
    onClick: _react.PropTypes.func,

    /**
     * A function to call when the checkbox is clicked. This
     * function will will be called with `(rowIndex, event)`. The
     * `TableBody` and `TableHeader` components will automatically
     * merge in a function to goggle the checkbox.
     */
    onCheckboxClick: _react.PropTypes.func,

    /**
     * A boolean if the row should automatically check all the `TableColumn`s in the row
     * and add the className `grow` to the one that is the biggest. You can also disable
     * individual columns by adding the className `.prevent-grow` to them.
     */
    autoAdjust: _react.PropTypes.bool.isRequired,

    /**
     * An optional function to call onMouseOver.
     */
    onMouseOver: _react.PropTypes.func,

    /**
     * An optional function to call onMouseLeave.
     */
    onMouseLeave: _react.PropTypes.func,

    /**
     * Boolean if the row is currently selected. If this value will be
     * injected by the `TableHeader` or `TableBody` component.
     */
    selected: _react.PropTypes.bool,

    /**
     * The row's index in the table. This is injected via the `TableBody` component.
     */
    index: _react.PropTypes.number
  };
  TableRow.defaultProps = {
    autoAdjust: true
  };
  TableRow.contextTypes = _headerContextTypes2.default;
  TableRow.childContextTypes = _rowContextTypes2.default;
  exports.default = TableRow;
});
